src/main/java/co/cask/gcp/bigquery/BigQueryExecute.java

```diff
@@ -0,0 +1,186 @@
+import com.google.cloud.bigquery.JobInfo;
+import com.google.cloud.bigquery.QueryJobConfiguration;
+import com.google.cloud.bigquery.TableId;
+import com.google.cloud.bigquery.TableResult;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.threeten.bp.Duration;
+
+import javax.annotation.Nullable;
+
+/**
+ * This class <code>BigQueryExecute</code> executes a single Cloud BigQuery SQL.
+ *
+ * The plugin provides the ability different options like choosing interactive or batch execution
+ * of sql query, setting of resulting dataset and table, enabling/disbaling cache, specifying whether
+ * the query being executed is legacy or standard and reattempt strategy.
+ */
+@Plugin(type = Action.PLUGIN_TYPE)
+@Name(BigQueryExecute.NAME)
+@Description("Execute a Google BigQuery SQL.")
+public final class BigQueryExecute extends Action {
+  private static final Logger LOG = LoggerFactory.getLogger(BigQueryExecute.class);
+  public static final String NAME = "BigQueryExecute";
+  private static final String RECORDS_OUT = "records.out";
+  private static final String MODE_BATCH = "batch";
+  private static final String CACHE_ENABLED = "enabled";
+  private static final String LEGACY_ENABLED = "legacy-enabled";
+
+  private Config config;
+
+  @Override
+  public void run(ActionContext context) throws Exception {
+    // Check if a SQL has been specified, if the SQL has not been specified,
+    // then there is nothing for this class to do.
+    if (config.sql == null || config.sql.isEmpty()) {
+      throw new IllegalArgumentException(
+        "SQL not specified. Please specify a SQL to execute"
+      );
+    }
+
+    QueryJobConfiguration.Builder builder = QueryJobConfiguration.newBuilder(config.sql);
+
+    // Run at batch priority, which won't count toward concurrent rate limit.
+    if (config.mode.equalsIgnoreCase(MODE_BATCH)) {
+      builder.setPriority(QueryJobConfiguration.Priority.BATCH);
+    } else {
+      builder.setPriority(QueryJobConfiguration.Priority.INTERACTIVE);
+    }
+
+    // Save the results of the query to a permanent table.
+    if (config.dataset != null && config.table != null
+      && !config.dataset.isEmpty() && !config.table.isEmpty()) {
+      builder.setDestinationTable(TableId.of(config.dataset, config.table));
+      context.getArguments().set(context.getStageName() + ".dataset", config.dataset);
+      context.getArguments().set(context.getStageName() + ".table", config.dataset);
+    }
+
+    // Enable or Disable the query cache to force live query evaluation.
+    if (config.cache.equalsIgnoreCase(CACHE_ENABLED)) {
+      builder.setUseQueryCache(true);
+    }
+
+    // Enable legacy SQL
+    if (config.legacy.equalsIgnoreCase(LEGACY_ENABLED)) {
+      builder.setUseLegacySql(true);
+    } else {
+      builder.setUseLegacySql(false);
+    }
+
+    QueryJobConfiguration queryConfig = builder.build();
+
+    // Location must match that of the dataset(s) referenced in the query.
```

Comment:
doesn't this mean the location must be a config property?

Reply:
fixed.